### ISSUE TITLE
Already generated PBKDF2 hashes not working with `mosquitto-go-auth`

### ISSUE THREAD

**Issue by magkopian** [iegomez/mosquitto-go-auth#239]
I have an existing database with PBKDF2 hashes which so far was being used with [mosquitto-auth-plug](https://github.com/jpmens/mosquitto-auth-plug) without a problem. The issue is that the hash checking by `mosquitto-go-auth` is done slightly differently even though I’m using the same settings.

Let me give you an example, let’s say I run the `pw` utility for the password "password" as follows.

`./pw -a sha256 -i 901 -h pbkdf2 -l 24 -s 12 -p "password"`

That will result to the following output,

`PBKDF2$sha256$901$1XIMQho85mmc6UIP$uSA2TA4QCGZl25+ze0aD7A1gtKbe19PL`

The code my existing hashes were generated was in PHP and is as follows,

```
<?php
$iterations = 901;
$algo = 'sha256';
$salt = base64_encode(openssl_random_pseudo_bytes(12));
$hash = base64_encode(hash_pbkdf2($algo, $mqtt_password, $salt, $iterations, 24, true));
$hashedPassword = sprintf("PBKDF2$%s$%d$%s$%s", $algo, $iterations, $salt, $hash);
```

Now, if I modify the above code slightly in order to generate the hash of the password "password" using the same salt that `pw` generated previously,

```
<?php
$iterations = 901;
$algo = 'sha256';
$salt = '1XIMQho85mmc6UIP';
$hash = base64_encode(hash_pbkdf2($algo, "password", $salt, $iterations, 24, true));
echo $hashedPassword = sprintf("PBKDF2$%s$%d$%s$%s", $algo, $iterations, $salt, $hash);
```

Then I get the following hash,

`PBKDF2$sha256$901$1XIMQho85mmc6UIP$bAYFlB2+kzd8j1JCm2sNAi7xqxpKiPpQ`

However, if I add a `base64_decode()` to the salt before I pass it to `hash_pbkdf2()`,

```
<?php
$iterations = 901;
$algo = 'sha256';
$salt = '1XIMQho85mmc6UIP';
$hash = base64_encode(hash_pbkdf2($algo, "password", base64_decode($salt), $iterations, 24, true));
echo $hashedPassword = sprintf("PBKDF2$%s$%d$%s$%s", $algo, $iterations, $salt, $hash);
```

Then I get the exact same hash that `pw` generated.

The problem, as I already explained, is that I already have hashes that were generated with the salt already being in base64 before the hash generation, which then fail authentication with  `mosquitto-go-auth` even though I’m using the correct configuration.

```
auth_opt_hasher pbkdf2
auth_opt_hasher_salt_size 12
auth_opt_hasher_iterations 901
auth_opt_hasher_keylen 24
auth_opt_hasher_algorithm sha256
```

By the way, I do know that is indeed the issue, because when I replaced the hash in my database with the one generated by `pw` I was able to authenticate without a problem.

_Basically, what I need is for `mosquitto-go-auth` to not base64 decode the salt from the hash in the database during the hash verification, before attempting to generate the hash using the password supplied by the client to compare it._

Unfortunately, I’m no Go programmer and I haven’t figured out how to modify your code in order to make the required changes before compiling the plugin. I would greatly appreciate it if you could help me with this.

**Comment by magkopian** [comment]
Ok, I'm an idiot. All I had to do was to set `auth_opt_hasher_salt_encoding` to `utf-8`. For some reason I completely missed that configuration option.

**Comment by iegomez** [comment]
Hehe, no worries, I'm glad you found it soon.

Cheers!

**Comment by iegomez** [comment]
I do see your PHP hash function is base64 encoding it: ` base64_encode(pbkdf2(...`.
Looks like it's just a matter of setting the encoding correctly in the plugin options.

**Comment by FernandoGarcia** [comment]
> I do see your PHP hash function is base64 encoding it: ` base64_encode(pbkdf2(...`. Looks like it's just a matter of setting the encoding correctly in the plugin options.

Hi!
Thanks for your answer.
In this way?
`auth_opt_mysql_salt_encoding base64`
Best regards.

**Comment by iegomez** [comment]
Please check https://github.com/iegomez/mosquitto-go-auth#hashing.
Notice that `base64` is already the default if omitted.

**Comment by FernandoGarcia** [comment]
Do you know why this password works even when the configuration above is enabled?

```
~/mosquitto-go-auth$ ./pw -p ec0245b
PBKDF2$sha512$100000$QlaICVjci0a43BGkDyAeaA==$d6XTMRduoO+nRHioE/t25e7STovcX6FZpHTabPTWTptMThzoZDe9ctyOtA1DsJC8PqI2NkMYWgII7mdfydvfBA==
```

**Comment by iegomez** [comment]
That should be a hint that something's misconfigured, but I can't tell with the provided information (haven't seen your whole conf, nor logs, etc.). I don't have the time to debug this either, but I'll recommend again that you set log level to debug and see what may be going on.

**Comment by FernandoGarcia** [comment]
Thanks for your time.

Unfortunately I can't see any usefull information at `/var/log/mosquitto/mosquitto.log` even debug level enabled.

Here you can my configuration and the result just changing the password directly in DB.

![Captura de tela de 2023-05-17 14 59 38](https://github.com/iegomez/mosquitto-go-auth/assets/2778501/bf4defb3-e1d3-4910-9177-dfa7fa43c415)

**Comment by AdamMiltonBarker** [comment]
> Thanks for your time.
> 
> Unfortunately I can't see any usefull information at `/var/log/mosquitto/mosquitto.log` even with debug level enabled.
> 
> Here you can my configuration and the result just changing the password directly in DB.
> 
> ![Captura de tela de 2023-05-17 14 59 38](https://github.com/iegomez/mosquitto-go-auth/assets/2778501/bf4defb3-e1d3-4910-9177-dfa7fa43c415)

Same issue here, debug outputs nothing other than not authorized.
